src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -710,58 +710,58 @@ private void pruneDeletedTombstones() {
         lastDeleteVersionPruneTimeMSec = timeMSec;
     }
 
-    // TODO: can we please remove this method?!
-    private void waitForMerges(boolean flushAfter, boolean upgrade) {
-        try {
-            Method method = IndexWriter.class.getDeclaredMethod("waitForMerges");
-            method.setAccessible(true);
-            method.invoke(indexWriter);
-        } catch (ReflectiveOperationException e) {
-            throw new OptimizeFailedEngineException(shardId, e);
-        }
-        if (flushAfter) {
-            flush(true, true, true);
-        }
-        if (upgrade) {
-            logger.info("Finished upgrade of " + shardId);
-        }
-    }
-
     @Override
     public void forceMerge(final boolean flush, int maxNumSegments, boolean onlyExpungeDeletes, final boolean upgrade) throws EngineException {
-        if (optimizeMutex.compareAndSet(false, true)) {
-            try (ReleasableLock lock = readLock.acquire()) {
-                ensureOpen();
-                /*
-                 * The way we implement upgrades is a bit hackish in the sense that we set an instance
-                 * variable and that this setting will thus apply to the next forced merge that will be run.
-                 * This is ok because (1) this is the only place we call forceMerge, (2) we have a single
-                 * thread for optimize, and the 'optimizeMutex' guarding this code, and (3) ConcurrentMergeScheduler
-                 * syncs calls to findForcedMerges.
-                 */
-                MergePolicy mp = indexWriter.getConfig().getMergePolicy();
-                assert mp instanceof ElasticsearchMergePolicy : "MergePolicy is " + mp.getClass().getName();
+        ensureOpen();
+        if (optimizeLock.tryLock()) { // we do NOT acquire the readlock here since we are waiting on the merges to finish - that's fine since the IW.rollback should interrupt all the
```

Comment:
Today if you call optimize while optimize is already running the 2nd one blocks right?  But now, the 2nd one will just return right away?